xdsl/dialects/csl/csl_stencil.py

```diff
@@ -121,6 +141,237 @@ def __init__(
+
+    num_chunks = prop_def(IntegerAttr)
+
+    res = var_result_def(stencil.TempType)
+
+    traits = frozenset([IsolatedFromAbove(), RecursiveMemoryEffect()])
+
+    def print(self, printer: Printer):
+        def print_arg(arg: tuple[SSAValue, Attribute]):
+            printer.print(arg[0])
+            printer.print(" : ")
+            printer.print(arg[1])
+
+        printer.print("(")
+
+        # args required by function signature, plus optional args for regions
+        args = [self.communicated_stencil, self.iter_arg, *self.args]
+
+        printer.print_list(
+            zip(args, (a.type for a in args)),
+            print_arg,
+        )
+        printer.print(") -> (")
+        printer.print_list((r.type for r in self.res), printer.print_attribute)
+        printer.print(") ")
+        printer.print_op_attributes(self.attributes, print_keyword=True)
+        printer.print("(")
+        printer.print_region(self.chunk_reduce, print_entry_block_args=True)
+        printer.print(", ")
+        printer.print_region(self.post_process, print_entry_block_args=True)
+        printer.print(")")
+
+    @classmethod
+    def parse(cls, parser: Parser):
+        def parse_args():
+            value = parser.parse_unresolved_operand()
+            parser.parse_punctuation(":")
+            type = parser.parse_attribute()
+            value = parser.resolve_operand(value, type)
+            return value
+
+        operands = parser.parse_comma_separated_list(parser.Delimiter.PAREN, parse_args)
+
+        props = parser.parse_optional_properties_dict()
+        topo = props["topo"] if "topo" in props else None
+        swaps = props["swaps"]
+        num_chunks = props["num_chunks"]
+
+        parser.parse_punctuation("->")
+        result_types = parser.parse_comma_separated_list(
+            parser.Delimiter.PAREN, parser.parse_attribute
+        )
+        attrs = parser.parse_optional_attr_dict_with_keyword()
+        if attrs is not None:
+            attrs = attrs.data
+        parser.parse_punctuation("(")
+        chunk_reduce = parser.parse_region()
+        parser.parse_punctuation(",")
+        post_process = parser.parse_region()
+        parser.parse_punctuation(")")
+        return cls(
+            operands=[operands[0], operands[1], operands[2:]],
+            result_types=[result_types],
+            regions=[chunk_reduce, post_process],
+            properties={"topo": topo, "swaps": swaps, "num_chunks": num_chunks},
+            attributes=attrs,
+        )
+
+    @staticmethod
+    def get(
+        args: Sequence[SSAValue] | Sequence[Operation],
+        body: Block | Region,
+        result_types: Sequence[stencil.TempType[Attribute]],
+    ):
+        assert len(result_types) > 0
+
+        if isinstance(body, Block):
+            body = Region(body)
+
+        return ApplyOp.build(
+            operands=[list(args)],
+            regions=[body],
+            result_types=[result_types],
+        )
```

Comment:
Can you make this an `__init__`?

Reply:
I probably want to take this out for now.